### ISSUE TITLE
NaN Exploits

### ISSUE THREAD

**Issue by Vladymyr96** [NoCheatPlus/Issues#333]
This can be used on a lot of servers (including serves with no cheat plus and another anticheats).
This puts ur vertical/horizontal location to NaN (Not a number), this can make you fly, die, teleport, noclip, god mode, glitch animation (your legs and arms doesn't move), freeze server exploit, nofall.

3.1.5.0-SNAPSHOT-sMD5NET-b1043

This always happen, sometimes you can die, you need to be horizontal collided 

It only works for 1.9.x - 1.11.x servers

The server i was testing on have a lot of anticheat but they are selectable

Links of the exploit:

       -NoClip/Fly: https://a.pomf.cat/tnzrja.mp4
       -NoClip: https://i.gyazo.com/94d9535103da3ec210abae7994fb1ef4.gif
       -Glitch Animation: https://i.gyazo.com/096c5c9a6b1391f09e2cb9708aa511c2.gif

Probably mojang did something wrong on the checks

**Comment by asofold** [comment]
Can they use this only as the first move after joining , or can they use it anywhere any time / at will?

**Comment by Vladymyr96** [comment]
They can use it any time, but the player must (vertical or horizontal) clip while colliding with a wall

**Comment by asofold** [comment]
A debug log would be nice to have (https://github.com/NoCheatPlus/Docs/wiki/Debugging#on-the-fly-debug-output-for-individual-players).

I'd assume they can somehow clip, e.g. with an untracked move. Do you know if they exploit the server teleporting players on such/similar occasions (Bukkit: teleport events with UNKNOWN reason, used to be uncancellable)?

**Comment by Vladymyr** [comment]
Sorry, I don't have access to the logs.

> Do you know if they exploit the server teleporting players on such/similar occasions?

I don't think so.

**Edit:** I took a look at NetHandlerPlayServer class and I saw **[this](http://hastebin.com/oquwafosig.cpp)** for CPacketPlayer, it doesn't have any checks for NaN. So the **[fix](http://hastebin.com/ecinoqikug.java)** would kick the player for "Invalid Position/Rotation" if something is NaN. It should solve the issue, right?

**Edit 2**: Okay, never mind they + and - NaN for the position.

**Comment by asofold** [comment]
Yes :), both is needed.

NoCheatPlus should sout out NaN for sure, but perhaps it's "too late" when NCP receives the events. Possibly the server will teleport them somehow, but i can't tell what's happening in between. We could monitor NaN and infinite on packet level, which probably is the best we can do (ProtocolLib).

**Comment by Vladymyr** [comment]
If you're not capable of fixing it, should I report this directly to Spigot?

**Comment by asofold** [comment]
In fact, this should probably be reported to spigot, especially considering the suggested fix!

(https://hub.spigotmc.org/jira/secure/Dashboard.jspa)

If you don't, i should do on monday.

**Comment by jonesdj1** [comment]
tried this out and set my position x to NaN and i fall in the void till i move my mouse but if you move your mouse just kills you after about 5 secs and I could not get it to work on my test server only worked a public test server ncp.b0at.xyz

**Comment by jonesdj1** [comment]
turns out this exploit could be caused from aac being installed on the same server most likely why it did not work on my test server.

**Comment by rbreslow** [comment]
This exploit is not reproducible on my server 2b2t.org. We're running the latest from GitHub on PaperSpigot. Also- my friend who runs Constantiam.net who is on the latest build of NCP on Spigot also cannot reproduce.

**Comment by KurtisAD** [comment]
I run the anarchy server Constantiam.net, and am not able to reproduce. Currently on the latest build of NCP and Spigot

**Comment by FirstReplay** [comment]
This issue IS happening on all servers including the ones above. The posters above have been using the exploit successfully on both servers. Latest NCP build and latest Paper build.

**Comment by rbreslow** [comment]
Excuse me?

I'm on the most recent branch (
https://github.com/NoCheatPlus/NoCheatPlus/commit/21eb53f7f2a35b3d57afce76002a4d03a638d258)
correct me if I'm wrong.

**Comment by Black-Hole** [comment]
JavaDoc for Doubles.isFinite() states:
Returns {@code true} if {@code value} represents a real number. This is equivalent to, but not necessarily implemented as, !(Double.isInfinite(value) || Double.isNaN(value)).

I suspect the exploit is using some other packet besides PacketPlayInFlying or it's subclasses.

**Comment by jonesdj1** [comment]
@Black-Hole does not use any packets it uses the bounding box and the exploit only works on some servers there for hard to reproduce everyone should update paperspigot and NCP and other plugins to fix as this is most likely something todo with plugins

**Comment by asofold** [comment]
Bounding box would mean .... pre 1.8 or so (unless indirectly causeing the server to set it oddly)?

For guessing, i could imagine attacks specific to protocol support plugins, combinations of anti cheating plugins, individual anti cheating plugins, ...

Perhaps there is some double arithmetics bomb somewhere with the vanilla checks (teleporting players out of blocks), or with another packet (VehicleSteer?). 

We do need to find out under which circumstances this is possible (no anti cheating plugins there, Vanilla server?, CRaftBukkit(Spigot), default Spigot ....). Alternatively a specific action/packet/content to check for ....

**Comment by jonesdj1** [comment]
@asofold this only works on some server aka i could only get it to work on a public ncp test server that is 1.11 that allows older versions to join but it is full 1.11 but no this would not work on 1.8 only 1.9 or higher but yeah you can set your bounding box like this Minecraft.getMinecraft().player.setEntityBoundingBox(AxisAlignedBB stuff here);

**Comment by Vladymyr96** [comment]
@jonesdj1 b0at is 1.10.2 but allows other versions.

**Comment by jonesdj1** [comment]
@Vladymyr96 he updated to 1.11 he had shulker boxes and other 1.11 only stuff

**Comment by asofold** [comment]
So client-side it works with bounding boxes, but they don't have a bounding box related packet on 1.11, do they?

**Comment by jonesdj1** [comment]
its not a packet you try to set your position with Minecraft.getMinecraft().player.setEntityBoundingBox();

**Comment by Vladymyr96** [comment]
@jonesdj1 ye like for example mc.player.setPosition(mc.player.posX + Double.NaN,  mc.player.posY + 0.001F, mc.player.posZ + Double.NaN);

**Comment by asofold** [comment]
I'd love to see which packets are used for bounding boxes...
pre-idk1.8-or-so, the moving packets could contain a stance value, but that has been removed. So what would it be now? Perhaps your'e getting kicked, because the hack  doesn't work server-side and you're floating too long, effectively? (i assume not)

Appart from the player position packets, which packets are there (on which version of MC)?

[bounding box?? / vehicle steer? vehicle move?]

**Comment by Vladymyr96** [comment]
@asofold I tested it on 1.11 and works for me.

**Comment by asofold** [comment]
setPosition would result in one of the ordinary player position packets - i can understand that if the sanity checking isn't done, it'll be able to create havoc. I was just asking for the bounding box adaption resulting in which packet. If it's adaptive, it could also just be a position packet.

@Vladymyr96  Why are you using 'mc.player.posX + Double.NaN', would  'Double.NaN' not work?

I have to review NCP, all comparison with == on Double.NaN (and possibly infinity too) is problematic. To avoid confusion, i'll route things through auxiliary methods, also where my own methods return Double.NaN and the like as primitive doubles.

**Comment by asofold** [comment]
Roughly at build 1049 ~ quick and dirty review of NaN and xyINFINITY use within NCP.

(Still doesn't explain why the server doesn't catch it early, since Doubles.isFinite should catch it. Thus it's probably not checked at all...)

**Comment by Xayanix** [comment]
I can confirm this issue. I am going to make public fix for it.
This is how error log looks like: http://pastebin.com/6yD5pJjp

**Comment by jonesdj1** [comment]
@Xayanix I dont think thats related what we are talking about :P

**Comment by Xayanix** [comment]
@jonesdj1 Yes, probably it is. Player joins server with NaN location and then server gets freezed.

Log before crash:

[18:44:08] [LoginProcessingThread/INFO]: UUID of player name_of_player is uuid_of_player
[18:44:08] [Server thread/INFO]: name_of_player[/ip:port] logged in with entity id 361724 at ([world]NaN, NaN, NaN)

**Comment by asofold** [comment]
The log by @Xayanix looks like it could be the issue. If that's the spot, then it looks like this should be reported to the jira at the developers hub of spigotmc.org. The log is from 1.10.2 though, it would be best to have such a log with the latest of spigot versions (MC 1.11.2). Reviewing the server code may reveal that 1.11 is still vulbnerable, i haven't checked.

(If i'm mistaken with reading from the log that plugins are not involved directly, please correct me.)

**Comment by Xayanix** [comment]
@asofold 
I updated server to 1.11.2 (2 days ago) and i have not reported any crashes. I will let you know if something will happen.

**Comment by jonesdj1** [comment]
yeah ive made my pos NaN NaN NaN in a player file on 1.11 and it was fine

**Comment by asofold** [comment]
@Xayanix Concerning the 1.10.2 log: I'd be somewhat surprised if that went through (NaN or infinity) with NCP build 1047 or later, with ProtocolLib and the FlyingFrequency check activated. In that case the packet should be received and cancelled by NCP (no kick, only a log message if the player is debugged).

**Comment by v0idst4r** [comment]
thanks Vlad for fixing my noclip hake

**Comment by Vladymyr96** [comment]
@v0idst4r it' got patche? yey and np bb <3 <3 <3

**Comment by v0idst4r** [comment]
It will soon becuz of you might as well break 1.9+ NCP flight as well

**Comment by jonesdj1** [comment]
mfw this was never a exploit nor could it be used

**Comment by v0idst4r** [comment]
I tried it and I effectively no clipped on cubecraft without being inside a
block

On Jan 11, 2017 11:24 PM, "jonesdj1" <notifications@github.com> wrote:

mfw this was never a exploit nor could it be used

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/NoCheatPlus/Issues/issues/333#issuecomment-272079252>,
or mute the thread

**Comment by jonesdj1** [comment]
@v0idst4r you can clip down but if you move your mouse you get teleported back lol

**Comment by v0idst4r** [comment]
Trust me, it works

**Comment by asofold** [comment]
@jonesdj1 You might have to wrap all player moving with the right sequence of packets, otherwise it'll be like turning around while standing in-air, which would get detected usually.


Can anybody still use this on a server that is having NCP build 1054 or later and ProtocolLib running?

**Comment by jonesdj1** [comment]
@asofold cubecraft uses a build of nocheatplus from like 6+ months ago or so without default config as people can fly on there with things that never would work on ncp.

**Comment by v0idst4r** [comment]
You can fly but it has to be either the glitch abusing 1.9+ NCP fly or you
have to place a block first

**Comment by asofold** [comment]
Latest dev counts, sorry :). For altering existing past-builds on random servers from the outer atmosphere i'd need a capable it-engineer and a much much better source of energy ...

@v0idst4r Is it related to NCP in particular? Does it work with NCP (Jenkins.-..)= build 1054?

**Comment by v0idst4r** [comment]
It's not a problem with NCP, I believe it's a problem with bukkit/spigot.
Normally sending a packet with NaN should kick you for "invalid move packet
received" but it doesn't on 1.9+ servers.

On Jan 16, 2017 5:40 PM, "asofold" <notifications@github.com> wrote:

Latest dev counts, sorry :). For altering existing past-builds on random
servers from the outer atmosphere i'd need a capable it-engineer and a much
much better source of energy ...

@v0idst4r <https://github.com/v0idst4r> Is it related to NCP in particular?
Does it work with NCP (Jenkins.-..)= build 1054?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/NoCheatPlus/Issues/issues/333#issuecomment-272986462>,
or mute the thread

**Comment by v0idst4r** [comment]
And as @Xayanix 's pastebin shows, you can even crash the server if you send just NaN and not your current position + NaN

**Comment by v0idst4r** [comment]
On latest 1.11.2 spigot this was fixed and now kicks the player for "Invalid move packet recieved"

**Comment by rbreslow** [comment]
Not fixed for my build of Spigot... I think this may be a problem with the
vanilla branch (from Mojang) and submitted a ticket to their Jira

**Comment by v0idst4r** [comment]
What build and MC version?

**Comment by asofold** [comment]
Yeah we'll need all the context(s)...
* Vanilla or Spigot (CraftBukkit?) + version.
* NCP version if used. [net checks activated?]
* ProtocolLib version if used.

Otherwise this is just confusing, i'd prefer not to have to write my own client right now (sending stuff via ProtocolLib might work, but it might not).

**Comment by jonesdj1** [comment]
@asofold seems to be fixed 1.11.2 PaperSpigot, Plugins: NoCheatPlus. Default Config's For Everything

**Comment by asofold** [comment]
If that applies, we'll need some means of mitigation for past versions though (1.9.x/1.10.x).

**Comment by v0idst4r** [comment]
Auto kick if a player's position isn't Double.isFinite ?

**Comment by asofold** [comment]
Well usually NCP should cancel such packets, but perhaps there is conditions when it doesn't happen on packet level:
* ProtocolLib not installed.
* The MovingFlying thing is not run for whatever reasons (flyingfrequency disabled and other).
* (Bugs?).

If NCP finds such a position in Bukkit it'd kick the player, but that's likely too late already, because the crashing and stuff seems to happen at another stage.

So perhaps i'll find the reason why the checks don't run.
* https://github.com/NoCheatPlus/NoCheatPlus/blob/master/NCPCompatProtocolLib/src/main/java/fr/neatmonster/nocheatplus/checks/net/protocollib/MovingFlying.java#L171
* https://github.com/NoCheatPlus/NoCheatPlus/blob/master/NCPCompatProtocolLib/src/main/java/fr/neatmonster/nocheatplus/checks/net/protocollib/ProtocolLibComponent.java#L89
(Quick glance: it should be checked and the packet cancelled? Does need ProtocolLib and not having deactivated net.flyingfrequency.)

**Comment by asofold** [comment]
So i'm still lacking the information, if these really work with NCP 1084 (or better latest) and ProtocolLib on (for the said past versions of Spigot). All moving checks and flyingfrequency / net activated of course.

**Comment by jonesdj1** [comment]
this was a bug on spigots end for a few 1.11 builds

**Comment by asofold** [comment]
@jonesdj1 Thanks :) - is that fixed for 1.9.4/1.10.2 too? Otherwise a fix would be nice to have.
